src/types/type_param.rs

```diff
@@ -24,15 +32,61 @@ pub enum TypeParam {
     ClassicType,
     /// Argument is a [TypeArg::HashableType]
     HashableType,
-    /// Node must provide a [TypeArg::List] (of whatever length)
-    /// TODO it'd be better to use [`Container`] here.
-    ///
-    /// [`Container`]: crate::types::simple::Container
-    List(Box<TypeParam>),
+    /// A nested definition containing other TypeParams (possibly including other [HashableType]s).
+    /// Note that if all components are [TypeParam::Value]s, then the entire [Container] should be stored
+    /// inside a [TypeParam::Value] instead; so there will never be a [Container::Opaque] here,
+    /// as there are no [CustomType]s corresponding to TypeParam's except Hashable [TypeParam::Value]s.
+    Container(Container<TypeParam>),
     /// Argument is a value of the specified type.
     Value(HashableType),
 }
 
+impl TypeParam {
+    fn value_types(
+        typarams: TypeRow<TypeParam>,
+    ) -> Result<TypeRow<HashableType>, TypeRow<TypeParam>> {
+        if typarams.iter().all(|e| matches!(e, TypeParam::Value(_))) {
+            Ok(typarams
+                .into_owned()
+                .into_iter()
+                .map(|e| match e {
+                    TypeParam::Value(ht) => ht,
+                    _ => panic!(), // We checked all matched above
+                })
+                .collect::<Vec<_>>()
+                .into())
+        } else {
+            Err(typarams)
+        }
+    }
+
+    /// New Tuple TypeParam, elements defined by TypeRow
+    pub fn new_tuple(elems: impl Into<TypeRow<TypeParam>>) -> Self {
+        match TypeParam::value_types(elems.into()) {
+            Ok(h_tys) => {
+                TypeParam::Value(HashableType::Container(Container::Tuple(Box::new(h_tys))))
+            }
+            Err(ty_params) => Self::Container(Container::Tuple(Box::new(ty_params))),
+        }
+    }
+
+    /// New Tuple typeparam, elements defined by TypeRow
+    pub fn new_sum(elems: impl Into<TypeRow<TypeParam>>) -> Self {
+        match TypeParam::value_types(elems.into()) {
+            Ok(h_tys) => TypeParam::Value(HashableType::Container(Container::Sum(Box::new(h_tys)))),
+            Err(ty_params) => Self::Container(Container::Sum(Box::new(ty_params))),
+        }
+    }
+
+    /// Creates a new TypeParam accepting values of a specified CustomType, which must be hashable.
```

Comment:
Hopefully the mere presence of this will prevent anyone creating a `TypeParam::Container(Container::Opaque(...))` directly